core/src/packets/icmp/v6/ndp/mod.rs

```diff
@@ -19,105 +19,797 @@
+        f.debug_struct("MutableNdpOption")
+            .field("option_type", &self.option_type())
+            .field("length", &self.length())
+            .field("$offset", &self.offset)
+            .field("$len", &(self.length() * 8))
+            .finish()
+    }
+}
+
+/// An iterator that iterates through the options in the NDP message body
+/// mutably.
+pub struct MutableNdpOptionsIterator<'a> {
+    mbuf: &'a mut Mbuf,
+    offset: usize,
+}
+
+impl MutableNdpOptionsIterator<'_> {
+    /// Advances the iterator and returns the next value.
+    ///
+    /// Returns `Ok(None)` when iteration is finished; returns `Err` when a
+    /// parse error is encountered during iteration.
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> Fallible<Option<MutableNdpOption<'_>>> {
+        if self.mbuf.data_len() > self.offset {
+            match MutableNdpOption::new(&mut self.mbuf, self.offset) {
+                Ok(option) => {
+                    // advances the offset to the next option
+                    self.offset = option.end_offset();
+                    Ok(Some(option))
+                }
+                Err(e) => Err(e),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl fmt::Debug for MutableNdpOptionsIterator<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MutableNdpOptionsIterator")
+            .field("offset", &self.offset)
+            .finish()
+    }
+}
+
+/// Options in the NDP message body.
+pub struct NdpOptions<'a> {
+    mbuf: &'a mut Mbuf,
+    offset: usize,
+}
+
+impl NdpOptions<'_> {
+    /// Returns an iterator that allows modifying each option.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// let advert = ipv6.parse::<RouterAdvertisement<Ipv6>>()?;
+    /// let mut options = advert.options_mut();
+    /// let mut iter = options.iter();
+    ///
+    /// while let Some(option) = iter.next()? {
+    ///     println!("{:?}", option);
+    /// }
+    /// ```
+    #[inline]
+    pub fn iter(&mut self) -> MutableNdpOptionsIterator<'_> {
+        MutableNdpOptionsIterator {
+            mbuf: self.mbuf,
+            offset: self.offset,
+        }
+    }
+
+    /// Prepends a new option `T` at the beginning of the options.
+    #[inline]
+    pub fn prepend<'a, T: NdpOption<'a>>(&'a mut self) -> Fallible<T> {
+        T::try_push(self.mbuf, self.offset, Internal(()))
+    }
+
+    /// Appends a new option `T` at the end of the options.
+    #[inline]
+    pub fn append<'a, T: NdpOption<'a>>(&'a mut self) -> Fallible<T> {
+        T::try_push(self.mbuf, self.mbuf.data_len(), Internal(()))
+    }
+
+    /// Retains only the options specified by the predicate.
+    ///
+    /// In other words, remove all options `o` such that f(o) returns false.
+    /// If an error occurs, all removals done prior to the error cannot be
+    /// undone.
+    pub fn retain<F>(&mut self, mut f: F) -> Fallible<()>
```

Comment:
I'd say, def example doc retain at least (if you don't want to for prepend/append).